### PR TITLE
Utilities for working with subsets of model components

### PR DESCRIPTION
**Pull request by Robbybp**
## Summary/Motivation:
Adds some utilities I have found useful when I want to work with a "subsystem" of variables and constraints. Typically I do this by specifying the components that I would like the subsystem to include or by specifying the components I would like the subsystem to exclude. In the former case I find it convenient to create a temporary block with references to the desired variables and constraints. In the latter case I find it convenient to fix and deactivate the variables and constraints that I do not want in the subsystem. I like doing this in a context manager so I don't forget to reset the components when I am done.

An example of this is to initialize a dynamic model by solving with inputs fixed.
```python
with TemporarySubsystemManager(to_fix=input_vars):
    solver.solve(model)
```

An example of constructing blocks, if strongly connected components were identified by e.g. block triangularization.
```python
for variables, constraints in strongly_connected_components:
    block = create_subsystem_block(constraints, variables)
    solver.solve(block)
```

These can be combined to, say, fix inputs then solve independent blocks
```python
with TemporarySubsystemManager(to_fix=input_vars):
    for variables, constraints in strongly_connected_components:
        block = create_subsystem_block(constraints, variables)
        solver.solve(block)
```

In addition, this PR adds a class that I've found useful for testing blocks of constraints that are meant to calculate certain variables, based on the value of certain other "input" variables. This class inherits from the `TemporarySubsystemManager` for fixing or deactivating, and allows iteration over a set of "input scenarios," in which the model can be solved and values compared to pre-specified output values for each scenario. I've found this useful for testing constraints that calculate thermodynamic properties over a range of input values.
```python
input_vars = [model.v1]
input_values = ComponentMap([(model.v1, [1.1, 2.1])])
output_values = ComponentMap([(model.v2, [1.2, 2.2])])
with ParamSweeper(n_scenario, input_values, output_values, to_fix=input_vars) as param_sweeper:
    for inputs, outputs in param_sweeper:
        solver.solve(model)
        for var, val in outputs.items():
            assert var.value == val
```

## Changes proposed in this PR:
- Add `create_subsystem_block` function, `TemporarySubsystemManager` class, and `ParamSweeper` class

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by michaelbynum** [comment]
I haven't had a chance to do a real review yet, but this looks awesome.

File: pyomo/util/subsystems.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    """ This class is a context manager for cases when we want to
+    temporarily fix or deactivate certain variables or constraints
+    in order to perform some solve or calculation with the resulting
+    subsystem.
+
+    We currently do not support fixing variables to particular values,
+    and do not restore values of variables fixed. This could change.
+    """
+
+    def __init__(self, to_fix=None, to_deactivate=None, to_reset=None):
```
Comment (blnicho) [inline pyomo/util/subsystems.py:51]: Could you expand the docstring for the class with the expected types for `to_fix`, `to_deactivate`, and `to_reset`?

File: pyomo/util/subsystems.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+#  ___________________________________________________________________________
+
+from pyomo.core.base.PyomoModel import ConcreteModel
+from pyomo.core.base.reference import Reference
+from pyomo.core.expr.visitor import identify_variables
+from pyomo.common.collections import ComponentSet, ComponentMap
+
+
+def create_subsystem_block(constraints, variables=None, include_fixed=False):
+    """ This function defines creates a block to serve as a subsystem with
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply pyomo/util/subsystems.py:18]: Fixed

File: pyomo/util/subsystems.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.core.base.PyomoModel import ConcreteModel
+from pyomo.core.base.reference import Reference
+from pyomo.core.expr.visitor import identify_variables
+from pyomo.common.collections import ComponentSet, ComponentMap
+
+
+def create_subsystem_block(constraints, variables=None, include_fixed=False):
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply pyomo/util/subsystems.py:17]: Done

File: pyomo/util/subsystems.py (hunk 1 above)
Comment (Robbybp) [reply pyomo/util/subsystems.py:51]: Updated

**Review by blnicho — APPROVED** [review]

**Pull request merged**